README.md
# project_pengaduan
project_pengaduan


change_log.txt
Version 1.0.0
Des-23-2022
bug fix : -

fitur : - membuat barcode
	- menggunakan excelreader
	- membuat qrcode
	- membuat pdf
	- autoload class

Version 1.0.1
Des-24-2022
bug fix : - lokasi file tidak ditemukan ketika menggunakan href atau src dan diarahkan ke folder assets pada sintax html

fitur : - tidak bisa mengakses file secara langsung dari url

Version 1.0.2
Des-26-2022
bug fix : -

fitur : - menggunakan nama class yang sama, tetapi pada folder berbeda

Version 1.0.3
Des-27-2022
bug fix : -

fitur : - update framework automatis sesuai dengan versi terbaru, folder dan file yang terdownload adalah folder bawaan dari framework ini

Version 1.0.4
Des-29-2022
bug fix : - url "not found" ketika menggunakan 4 parameter pada url, contoh : https://domain.com/param_1/param_2/param_3/param_4
          - class database dan class awal tidak ikut terdownload ketika pertama mendownload framework        

fitur : -

Versi 1.0.5
Jan-13-2023
bug fix : - tidak dapat membedakan protocol BASEURL ketika website online tetapi menggunakan http

fitur : - pengaturan database terdapat pada file " pengaturan.php "
        - dapat membedakan pengaturan database online dan offline
        - autoupdate dipindah ke dalam file pengaturan.php
        - untuk penggunaan online, tidak perlu mengatur nilai dari variable $_PENGATURAN['ROOR_DIR']

Versi 1.0.6
Jan-15-2023
bug fix : - setelah dilakukan update, file bawaan saat mendownload pertama kali, masih terikut

fitur : - update automatis ketika versi dari developer berubah
        
Versi 1.0.7
Feb-19-2023
bug fix : - parameter ke #2 dan #3 pada url tidak bisa menggunakan angka dan huruf, atau gabungan dari keduanya
            contoh : https://domain.com/param_1/param_2/param_3/param_4
            
fitur : - tidak mengizinkan mengambil isi content html menggunakan file_get_contents dari php
        - menambahkan keamanan untuk request cross domain, pada file keamanan.php
        
Versi 1.0.8
Mar-02-2023
bug fix : - tidak dapat menampilkan file error 404 ketika url tidak ditemukan

fitur : - membuat list request yang akan diamankan ketika request cross domain
        - menambahkan custom message untuk request cross domain yang diamankan
        
Versi 1.0.9
Mar-09-2023
bug fix : - ketika pindah halaman selalu ada slash pada akhir url

fitur : - mengoptimalkan ukuran transfer html file dengan menghilangkan komentar dari html
        - menambahkan pilihan slash atau tidak ketika menggunakan $this->view->pindah($url, $slash);
          nilai default $slash : true
          contoh :
                  - ketika tidak ingin ada slash pada akhir url, gunakan  $this->view->pindah('url/yang_dituju', false);

Versi 1.1.0
Mar-15-2023
bug fix : - tidak bisa menjalankan javascript ketika javascript di komentar

fitur : -

Versi 1.1.1
Mar-16-2023
bug fix : -

fitur : - mengambil isi file view sebagai text

Versi 1.1.2
Mar-22-2023
bug fix : -

fitur : - menambahkan template engine untuk memudahkan pengaturan file

Versi 1.1.3
Mar-22-2023
bug fix : - tidak bisa menampilkan halaman ketika menggunakan fungsi view->show('nama_file');
          
fitur : -

Versi 1.1.4
Mar-23-2023
bug fix : - tidak dapat mengoptimalkan ukuran transfer file dari versi 1.0.9

fitur : -

Versi 1.1.5
Mar-30-2023
Bug fix : - tidak dapat mengoptimalkan besaran ukuran transfer file ketika menggunakan template
          - akses tidak diizinkan ketika menggunakan ip versi 6 pada $_PENGATURAN['SECURE_CROSS']

fitur : - validasi ip versi 4 dan ip versi 6 ketika mengakses web

Versi 1.1.6
Apr-25-2023
bug fix : - 

fitur : - membuat koneksi ke lebih dari 1 database

Versi 1.1.7
Mei-12-2023
bug fix : - untuk php versi 8.2 keatas, aktifkan extension "php_zip.dll" pada file "php.ini" untuk menggunakan class ZipArchive dari php

fitur : -

Version 1.1.8
Juni-4-2023
bug fix : - tidak dapat mengextract Zip update file

fitur : - menambahkan beberapa fungsi untuk class mysql

Version 1.1.9
Juni-5-2023
bug fix : -

fitur : - menambahkan file " memory_usage " untuk monitoring penggunaan RAM ketika mengakses website

Version 1.2.0
Juni-8-2023
bug fix : - tidak bisa akses ke subfolder untuk folder assets/css | assets/js | assets/img 

fitur : - 

Version 1.2.1
Juni-21-2023
bug fix : - tidak dapat mengekstrak file update ketika selesai terdownload updatean terbaru

fitur : -

Versi 1.2.2
Juni-23-2023
bug fix : -

fitur : - menghilangkan komentar dari html dan javascript ketika ditampilkan

Versi 1.2.3
Juni-25-2023
bug fix : -

fitur : - mengarahkan semua request/permintaan ke file index.php

Versi 1.2.4
Juni-26-2023
bug fix : -

fitur : - menambahkan fitur untuk simulasi halaman / tampilan.
         - file simulasi_url akan dibuat ketika file tersebut tidak ada
         - dokumentasi terkait fitur simulasi halaman berada pada file simulasi_url
         
Versi 1.2.5
Juni-29-2023
bug fix : - file assets dalam folder font | QrData yang tidak dapat diakses
          - tidak dapat mengakses subfolder dalam assets ketika subfolder lebih dari 1
fitur : -

Versi 1.2.6
Juli-11-2023
bug fix : - file asset css, js, dan img tidak ditemukan ketika menggunakan simulasi_url

fitur : -

Versi 1.2.7
Juli-31-2023
bug fix : - memperbaiki path untuk pengambilan assets ketika ingin menggunakan folder tersendiri

fitur : - menambahkan file cache untuk halaman yang akan ditampilkan
        - menambahkan $_PENGATURAN['ALLOWED_ASSETS'] = ['css', 'js', 'img', 'font', 'doc', 'QrData'];
          pada bagian akhir dari file pengaturan.php

Versi 1.2.8
Agustus-03-2023
bug fix : -

fitur : - bisa upload file dari local ke domain online sesuai dengan domain yang di tulis
          pada .online_update
        - untuk lebih jelasnya lihat dokumentasi pada file .online_update

Versi 1.2.9
Agustus-04-2023
bug fix : - tidak dapat mengupload file menggunakan .online_update jika file tersebut 
            lebih dari 1

fitur : - 

Versi 1.3.0
Agustus-05-2023
bug fix : -

fitur : - bisa download file dari domain online ke locak menggunakan .online_update


page/view.web/main.view.php
<header class="header-2">
    <div class="page-header section-height-75 relative" style="background-image: url('https://www.paluposo.id/wp-content/uploads/2023/01/WhatsApp-Image-2023-01-25-at-15.59.15.jpeg')">
      <div class="container">
        <div class="row">
          <div class="col-lg-7 text-center mx-auto" style="padding-top: 5rem !important;">
            <h1 class="text-white pt-3 mt-n5" style="font-size: 2.2rem !important;">Sistem Aplikasi Lapor Keluhan Masyrakat Kelurahan Tanamodindi Kota Palu</h1>
            <p class="lead text-white mt-3">Layanan Aspirasi dan Pengaduan Online Rakyat. <br /> Sampaikan laporan Anda langsung kepada instansi pemerintah berwenang. </p>
          </div>
        </div>
      </div>
      <div class="position-absolute w-100 z-index-1 bottom-0">
        <svg class="waves" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 24 150 40" preserveAspectRatio="none" shape-rendering="auto">
          <defs>
            <path id="gentle-wave" d="M-160 44c30 0 58-18 88-18s 58 18 88 18 58-18 88-18 58 18 88 18 v44h-352z" />
          </defs>
          <g class="moving-waves">
            <use xlink:href="#gentle-wave" x="48" y="-1" fill="rgba(255,255,255,0.40" />
            <use xlink:href="#gentle-wave" x="48" y="3" fill="rgba(255,255,255,0.35)" />
            <use xlink:href="#gentle-wave" x="48" y="5" fill="rgba(255,255,255,0.25)" />
            <use xlink:href="#gentle-wave" x="48" y="8" fill="rgba(255,255,255,0.20)" />
            <use xlink:href="#gentle-wave" x="48" y="13" fill="rgba(255,255,255,0.15)" />
            <use xlink:href="#gentle-wave" x="48" y="16" fill="rgba(255,255,255,0.95" />
          </g>
        </svg>
      </div>
    </div>
  </header>

  <section class="my-1 py-0">
    <div class="container">
      <div class="row align-items-center">
        <div class="col-lg-6 w-100 ">
          <div class="row justify-content-center">
            <div class="col-md-6 w-100">
              <div >
                <div class="card-body" style="text-align: center;">
                  <button class="btn btn-icon btn-3 btn-primary" type="button" onclick="kirimgambar(this)" >
                    <span class="btn-inner--icon"><i class="ni ni-button-play"></i></span>
                    <span class="btn-inner--text">Buat Laporan !!</span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>


  <section class="my-1 py-0" id="isikirimgambar" style="display: none;">
    <div class="container">
      <div class="row align-items-center">
        <div class="col-lg-6 w-100 ">
          <div class="row justify-content-center">
            <div class="col-md-6 w-100">
              <div style="box-shadow: 0 20px 27px 0 rgb(0 0 0 / 71%);border-radius: 40px;">
                <div class="card-body">
                  <h3 class="card-title text-info text-gradient" style="text-align:center;">Sampaikan Laporan Anda</h3>
                  <button class="btn btn-icon-only btn-2 btn-primary" type="button">
                    <span class="btn-inner--icon"><i class="ni ni-bulb-61"></i></span>
                  </button>
                  <blockquote class="blockquote text-white mb-0">
                    <form id="laporankuuyy" onsubmit="return false">


                      <div class="form-group ms-3">
                        <div class="input-group">
                          <select class="form-control" name="Ikategori" id="Ikategori" style="background-color: white;"> 
                            <option>--Pilih Kategori--</option>
                            <?php while ($a = $data['kat']->fetch_assoc()) { ?>
                              <option><?=$a['kategori']?></option>
                            <?php } ?>
                          </select>
                        </div>
                      </div>

                      <div class="form-group ms-3">
                        <div class="input-group">
                          <input class="form-control" placeholder="Masukan Nama Anda"  type="text" name="Inama" id="Inama">
                        </div>
                      </div>

                      <div class="form-group ms-3">
                        <div class="input-group">
                          <input class="form-control" placeholder="No hp" type="text" name="Inomor" id="Inomor">
                        </div>
                      </div>

                      <div class="form-group ms-3">
                        <div class="input-group">
                          <select class="form-control" name="Ikelurahan" id="Ikelurahan" style="background-color: white;"> 
                            <option selected>Tanamodindi</option>
                          </select>
                        </div>
                      </div>

                      <div class="form-group ms-3">
                        <div class="input-group">
                          <select class="form-control" style="background-color: white;" 
                          onchange="pilihrt(this)" name="Irw" id="Irw"> 
                          <option selected disabled>--Pilih Rw--</option>
                          <?php while ($b =$data['rw']->fetch_assoc()) { ?>
                            <option><?=$b['rw']?></option>
                          <?php } ?>
                        </select>
                      </div>
                    </div>

                    <div class="form-group ms-3">
                      <div class="input-group">
                        <select class="form-control" name="Irt" id="Irt" style="background-color: white;"> 
                          <option selected disabled>--Pilih Rt--</option>
                        </select>
                      </div>
                    </div>

                    <div class="form-group ms-3">
                      <div class="input-group">
                        <textarea name="Idesk" id="Idesk" rows="6" class="form-control textarea-flex autosize" placeholder="Ketik Isi Laporan Anda *" required="" style="overflow: hidden; overflow-wrap: break-word; height: 157.333px;"></textarea>
                      </div>
                    </div>

                    <div class="form-group ms-3">
                      <div class="input-group">
                        <div class="fallback">
                          <input name="fotolaporankuy[]"  type="file" class="form-control dropzone"  multiple />
                        </div>
                      </div>
                    </div>


                    <footer class=" text-info text-sm ms-3">
                      <hr class="horizontal dark mb-5">
                      <div class="container">
                        <div class="row">
                          <div class="col-12">
                            <button type="button" class="btn bg-gradient-primary mb-0 w-100" onclick="kirimlaporanku(this)" >Kirim Pengaduan !! </button>
                          </div>
                        </div>
                      </div>
                    </footer>
                  </blockquote>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>  




  <!-- -------- START Features w/ title and 3 infos -------- -->
  <section class="py-6" id="dataku">
    <div class="container">
      <div class="row">
        <div class="col-lg-6 mx-auto text-center pb-4">
          <h4 class="text-gradient text-primary mb-5">DataLaporan</h4>

          <!-- <h2>Apa Itu !?</h2> -->
          <!-- <p class="lead">Penyelenggara dapat mengelola pengaduan dari masyarakat secara sederhana, cepat, tepat, tuntas, dan terkoordinasi dengan baik;
            Penyelenggara memberikan akses untuk partisipasi masyarakat dalam menyampaikan pengaduan; dan
          Meningkatkan kualitas pelayanan publik.</p> -->
        </div>
      </div>
      <div class="row">
        <div class="col-lg-3 col-md-6">
          <div class="p-3 text-center">
            <div class="icon icon-shape bg-gradient-primary shadow mx-auto">
              <i class="ni ni-email-83"></i>
            </div>
            <h5 class="mt-4">Laporan Masuk</h5>

            <p><?=$data['pmsk']?></p>
          </div>
        </div>
        <div class="col-lg-3 col-md-6">
          <div class="p-3 text-center">
            <div class="icon icon-shape bg-gradient-warning shadow mx-auto">
              <i class="ni ni-email-83"></i>
            </div>
            <h5 class="mt-4">Laporan Pending</h5>
            <p><?=$data['pm']?></p>
          </div>
        </div>
        <div class="col-lg-3 col-md-6">
          <div class="p-3 text-center">
            <div class="icon icon-shape bg-gradient-danger shadow mx-auto">
              <i class="ni ni-email-83"></i>
            </div>
            <h5 class="mt-4">Laporan Diproses</h5>
            <p><?=$data['pp']?></p>
          </div>
        </div>
        <div class="col-lg-3 col-md-6 mx-md-auto">
          <div class="p-3 text-center">
            <div class="icon icon-shape bg-gradient-success shadow mx-auto">
             <i class="ni ni-email-83"></i>
           </div>
           <h5 class="mt-4">Laporan Selesai</h5>
           <p><?=$data['ps']?></p>
         </div>
       </div>
     </div>
   </div>
 </section>
 <!-- -------- END Features w/ title and 3 infos -------- -->

 <!-- END Section Content -->
 <!-- -------   START PRE-FOOTER 2 - simple social line w/ title & 3 buttons    -------- -->
 <!-- <div class="pt-5" id="documentku">
  <div class="container">
    <div class="row">
      <div class="col-lg-5 ms-auto">
        <h4 class="mb-1">Thank you for your support!</h4>
        <p class="lead mb-0">We deliver the best web products</p>
      </div>
      <div class="col-lg-5 me-lg-auto my-lg-auto text-lg-right mt-5">
        <a href="https://twitter.com/intent/tweet?text=Check%20Soft%20UI%20Design%20System%20made%20by%20%40CreativeTim%20%23webdesign%20%23designsystem%20%23bootstrap5&url=https%3A%2F%2Fwww.creative-tim.com%2Fproduct%2Fsoft-ui-design-system" class="btn btn-info mb-0 me-2" target="_blank">
          <i class="fab fa-twitter me-1"></i> Tweet
        </a>
        <a href="https://www.facebook.com/sharer/sharer.php?u=https://www.creative-tim.com/product/soft-ui-design-system" class="btn btn-primary mb-0 me-2" target="_blank">
          <i class="fab fa-facebook-square me-1"></i> Share
        </a>
        <a href="https://www.pinterest.com/pin/create/button/?url=https://www.creative-tim.com/product/soft-ui-design-system" class="btn btn-dark mb-0 me-2" target="_blank">
          <i class="fab fa-pinterest me-1"></i> Pin it
        </a>
      </div>
    </div>
  </div>
</div> -->
<!-- -------   END PRE-FOOTER 2 - simple social line w/ title & 3 buttons    -------- -->

<script type="text/javascript">
function formatPhoneNumber(phoneNumber) {
  const cleanedNumber = phoneNumber.replace(/[^\d]/g, '');
  const match = cleanedNumber.match(/^(0\d{3})(\d{4})(\d{4})$/);
  if (match) {
    return `${match[1]}-${match[2]}-${match[3]}`;
  }
  return phoneNumber;
}

const phoneInput = document.getElementById('Inomor');
phoneInput.addEventListener('input', function () {
  const rawValue = this.value;
  const formattedValue = formatPhoneNumber(rawValue);
  this.value = formattedValue;
});

phoneInput.addEventListener('keypress', function (event) {
  const keyCode = event.which ? event.which : event.keyCode;
  if (keyCode < 48 || keyCode > 57) {
    event.preventDefault();
  }
});

function showErrorAlert() {
  Swal.fire({
    icon: 'error',
    title: 'Format Nomor Salah',
    text: 'Mohon masukkan nomor telepon dengan format yang benar.',
    confirmButtonColor: '#d33',
    confirmButtonText: 'OK'
  });
}

phoneInput.addEventListener('blur', function () {
  const cleanedNumber = this.value.replace(/[^\d]/g, '');
  const match = cleanedNumber.match(/^(0\d{3})(\d{4})(\d{4})$/);
  if (!match) {
    showErrorAlert();
    this.focus();
  }
});

</script>